src/thor/triplegbuilder.cc

```diff
@@ -91,93 +91,190 @@ void AssignAdmins(const AttributesController& controller,
+                               itr == cuts.end() ? speed : itr->speed,
+                               itr == cuts.end() ? UNKNOWN_CONGESTION_VAL : itr->congestion,
+                               {&i}});
       }
+    }
+  }
 
-      // Set shape attributes speed per shape point if requested
-      if (controller.attributes.at(kShapeAttributesSpeed)) {
-        // convert speed to decimeters per sec and then round to an integer
-        trip_path.mutable_shape_attributes()->add_speed((distance * kDecimeterPerMeter / time) + 0.5);
-      }
+  // Find the first cut to the right of where we start on this edge
+  double distance_total_pct = src_pct;
+  auto cut_itr = std::find_if(cuts.cbegin(), cuts.cend(),
+                              [distance_total_pct](const decltype(cuts)::value_type& s) {
+                                return distance_total_pct <= s.percent_along;
+                              });
+  assert(cut_itr != cuts.cend());
+
+  // Set the shape attributes
+  for (auto i = shape_begin + 1; i < shape.size(); ++i) {
+    // when speed changes we need to make a new shape point and continue from there
+    double distance = shape[i].Distance(shape[i - 1]); // meters
+    double distance_pct = distance / edge->length();
+    double next_total = distance_total_pct + distance_pct;
+    size_t shift = 0;
+    if (next_total > cut_itr->percent_along && std::next(cut_itr) != cuts.cend()) {
+      // Calculate where the cut point should be between these two existing shape points
+      auto coef = (cut_itr->percent_along - distance_total_pct) / (next_total - distance_total_pct);
+      auto point = shape[i - 1].PointAlongSegment(shape[i], coef);
+      shape.insert(shape.begin() + i, point);
+      next_total = cut_itr->percent_along;
+      distance *= coef;
+      shift = 1;
+    }
+    distance_total_pct = next_total;
+    double time = distance / cut_itr->speed; // seconds
+
+    // Set shape attributes time per shape point if requested
+    if (controller.attributes.at(kShapeAttributesTime)) {
+      // convert time to milliseconds and then round to an integer
+      leg.mutable_shape_attributes()->add_time((time * kMillisecondPerSec) + 0.5);
+    }
+
+    // Set shape attributes length per shape point if requested
+    if (controller.attributes.at(kShapeAttributesLength)) {
+      // convert length to decimeters and then round to an integer
+      leg.mutable_shape_attributes()->add_length((distance * kDecimeterPerMeter) + 0.5);
+    }
 
-      // If we just cut the shape we need to go on to the next marker only after setting the attribs
-      std::advance(speed_itr, shift);
+    // Set shape attributes speed per shape point if requested
+    if (controller.attributes.at(kShapeAttributesSpeed)) {
+      // convert speed to decimeters per sec and then round to an integer
+      leg.mutable_shape_attributes()->add_speed((distance * kDecimeterPerMeter / time) + 0.5);
     }
+
+    // Set the incidents if we just cut or we are at the end
```

Comment:
this loop in the AddShapeAttributes function is what does the cutting and setting of attributes. The only real difference is here where we add incidents to the legs.